File: development/scripts/clean_workspace.py
import os
import shutil
import platform

def remove_dir(path):
    if os.path.exists(path):
        print(f"Removing directory: {path}")
        shutil.rmtree(path)

def remove_file(path):
    if os.path.exists(path):
        print(f"Removing file: {path}")
        os.remove(path)

def clean_workspace():
    # Clean yarn cache
    print("Cleaning yarn cache...")
    os.system("yarn cache clean")

    # Root directory cleanup
    remove_dir("node_modules")
    remove_dir(".expo")
    remove_dir(".husky/_")
    remove_dir(".app-mono-ts-cache")

    # Desktop cleanup
    remove_dir("apps/desktop/node_modules")
    remove_dir("apps/desktop/.expo")
    remove_dir("apps/desktop/__generated__")
    remove_dir("apps/desktop/dist")
    remove_dir("apps/desktop/build")
    remove_dir("apps/desktop/build-electron")
    remove_dir("apps/desktop/public/static/js-sdk")
    remove_dir("apps/desktop/public/static/connect")
    remove_file("apps/desktop/public/static/preload.js")

    # Ext cleanup
    remove_dir("apps/ext/node_modules")
    remove_dir("apps/ext/.expo")
    remove_dir("apps/ext/build")
    remove_file("apps/ext/src/entry/injected.js")
    remove_file("apps/ext/src/entry/injected.text-js")

    # Mobile cleanup
    remove_dir("apps/mobile/node_modules")
    remove_dir("apps/mobile/.expo")
    remove_dir("apps/mobile/__generated__")
    remove_dir("apps/mobile/ios/Pods")
    remove_dir("apps/mobile/ios/build")
    remove_dir("apps/mobile/ios/OneKeyWallet/web-embed")
    remove_dir("apps/mobile/ios/OneKeyWallet.xcworkspace/xcuserdata")
    remove_dir("apps/mobile/src/public/static/connect")
    remove_dir("apps/mobile/android/.gradle")
    remove_dir("apps/mobile/android/build")
    remove_dir("apps/mobile/android/app/build")
    remove_dir("apps/mobile/android/lib-keys-secret/build")
    remove_dir("apps/mobile/android/lib-keys-secret/.cxx")
    remove_dir("apps/mobile/android/app/src/main/assets/web-embed")

    # Web cleanup
    remove_dir("apps/web/node_modules")
    remove_dir("apps/web/.expo")
    remove_dir("apps/web/__generated__")
    remove_dir("apps/web/dist")
    remove_dir("apps/web/web-build")
    remove_dir("apps/web/.expo-shared")

    # Web-embed cleanup
    remove_dir("apps/web-embed/node_modules")
    remove_dir("apps/web-embed/.expo")
    remove_dir("apps/web-embed/__generated__")
    remove_dir("apps/web-embed/dist")
    remove_dir("apps/web-embed/web-build")
    remove_dir("apps/web-embed/.expo-shared")

    # Package cleanup
    remove_dir("packages/components/node_modules")
    remove_dir("packages/core/node_modules")
    remove_dir("packages/kit/node_modules")
    remove_file("packages/kit/src/components/WebView/injectedNative.text-js")
    remove_dir("packages/kit-bg/node_modules")
    remove_dir("packages/shared/node_modules")
    remove_file("packages/shared/src/web/index.html")

if __name__ == "__main__":
    clean_workspace()
    print("Workspace cleaned.")

File: development/scripts/clean_workspace.js
/* cspell:words prebundle */
const { execSync } = require('child_process');
const fs = require('fs');
const os = require('os');
const path = require('path');

const {
  getMobileShellCacheRoot,
  getSharedCacheRoot,
} = require('../../apps/mobile/scripts/dev-cache-paths');

// Function to remove directory recursively
function removeDir(dir, log) {
  if (fs.existsSync(dir)) {
    fs.rmSync(dir, { recursive: true, force: true });
    log(`Removed: ${dir}`);
  }
}

// Define directories to remove
const dirsToRemove = [
  // root
  './node_modules',
  './.expo',
  './.husky/_',
  './.app-mono-ts-cache',

  // desktop
  './apps/desktop/node_modules',
  './apps/desktop/.expo',
  './apps/desktop/__generated__',
  './apps/desktop/dist',
  './apps/desktop/build',
  './apps/desktop/build-electron',
  './apps/desktop/public/static/js-sdk',
  './apps/desktop/public/static/connect',
  './apps/desktop/public/static/preload.js',

  // ext
  './apps/ext/node_modules',
  './apps/ext/.expo',
  './apps/ext/build',
  './apps/ext/src/entry/injected.js',
  './apps/ext/src/entry/injected.text-js',

  // mobile
  './apps/mobile/node_modules',
  './apps/mobile/.expo',
  './apps/mobile/__generated__',
  './apps/mobile/out-dir-bundle',
  './apps/mobile/ios/Pods',
  './apps/mobile/ios/build',
  './apps/mobile/ios/outputs',
  './apps/mobile/ios/OneKeyWallet/web-embed',
  './apps/mobile/ios/OneKeyWallet.xcworkspace/xcuserdata',
  './apps/mobile/src/public/static/connect',
  './apps/mobile/android/.gradle',
  './apps/mobile/android/build',
  './apps/mobile/android/app/build',
  './apps/mobile/android/lib-keys-secret/build',
  './apps/mobile/android/lib-keys-secret/.cxx',
  './apps/mobile/android/app/src/main/assets/web-embed',

  // web
  './apps/web/node_modules',
  './apps/web/.expo',
  './apps/web/__generated__',
  './apps/web/dist',
  './apps/web/web-build',
  './apps/web/.expo-shared',

  // web-embed
  './apps/web-embed/node_modules',
  './apps/web-embed/.expo',
  './apps/web-embed/__generated__',
  './apps/web-embed/out-dir-bundle',
  './apps/web-embed/dist',
  './apps/web-embed/web-build',
  './apps/web-embed/.expo-shared',

  // components
  './packages/components/node_modules',

  // core
  './packages/core/node_modules',

  // kit
  './packages/kit/node_modules',
  './packages/kit/src/components/WebView/injectedNative.text-js',
  './packages/kit/src/components/WebView/injectedNative.js.txt',
  './packages/kit/src/components/WebViewWebEmbed/injectedWebEmbed.text-js',
  './packages/kit/src/components/WebViewWebEmbed/injectedWebEmbed.js.LICENSE.txt',
  './packages/kit/src/components/WebView/translateInject.text-js',
  './packages/kit/src/components/LightweightChart/utils/lightweightChartsStandalone.text-js',

  // kit-bg
  './packages/kit-bg/node_modules',
  './packages/kit-bg/src/desktopApis/injectedDesktopCode.text-js',

  // shared
  './packages/shared/node_modules',
  './packages/shared/src/web/index.html',
];

function cleanWorkspace({
  repoRoot = path.resolve(__dirname, '../..'),
  env = process.env,
  platform = process.platform,
  homeDirectory = os.homedir(),
  run = execSync,
  log = console.log,
} = {}) {
  log('Cleaning workspace...');
  log('Cleaning yarn cache...');
  run('yarn cache clean', { stdio: 'inherit', cwd: repoRoot, env });

  const sharedRoots = new Set([
    getMobileShellCacheRoot({}, homeDirectory),
    getMobileShellCacheRoot(env, homeDirectory),
    getSharedCacheRoot({}, platform, homeDirectory),
    getSharedCacheRoot(
      { ...env, ONEKEY_METRO_PREBUNDLE_CACHE_DIR: undefined },
      platform,
      homeDirectory,
    ),
    getSharedCacheRoot(env, platform, homeDirectory),
  ]);
  for (const root of sharedRoots) {
    if (fs.existsSync(root)) {
      // Custom roots may contain unrelated files; OneKey artifacts live in version directories.
      for (const entry of fs.readdirSync(root, { withFileTypes: true })) {
        if (
          /^v\d+$/u.test(entry.name) &&
          (entry.isDirectory() || entry.isSymbolicLink())
        ) {
          removeDir(path.join(root, entry.name), log);
        }
      }
    }
  }
  for (const dir of dirsToRemove) removeDir(path.resolve(repoRoot, dir), log);
  log('Workspace cleaned successfully.');
}

if (require.main === module) cleanWorkspace();

module.exports = { cleanWorkspace };
